javascript/ql/src/semmle/javascript/frameworks/WebSocket.qll

```diff
@@ -18,26 +18,64 @@ import javascript
  */
 private string channelName() { result = "message" }
 
+/**
+ * The names of the libraries modelled in this file.
+ */
+private module LibraryNames {
+  string sockjs() { result = "SockJS" }
+
+  string websocket() { result = "WebSocket" }
+
+  string ws() { result = "ws" }
+
+  class LibraryName extends string {
+    LibraryName() {
+      this = sockjs() or this = websocket() or this = ws()
+    }
+  }
+}
+
+/**
+ * Holds if the websocket library named `client` can send a message to the library named `server`.
+ * Both `client` and `server` are library names defined in `LibraryNames`.
+ */
+private predicate areLibrariesCompatible(LibraryNames::LibraryName client, LibraryNames::LibraryName server) {
+  // sockjs is a WebSocket emulating library, but not actually an implementation of WebSockets.
+  client = LibraryNames::sockjs() and server = LibraryNames::sockjs()
+  or
+  server = LibraryNames::ws() and
+  (client = LibraryNames::ws() or client = LibraryNames::websocket())
```

Comment:
checking: there's no server for `websocket()`?

Reply:
Correct.
[WebSocket](https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API) is a client-only (and browser only) WebSocket implementation.